src/lib/device-state.ts

```diff
@@ -47,6 +47,35 @@ export const serviceInstallFromImage = (
 export const formatImageLocation = (imageLocation: string) =>
 	imageLocation.toLowerCase();
 
+export type DeviceConfigHook = (
+	configVars: Dictionary<string>,
+	osVersion: string,
+) => void;
+
+export interface HookRegistry {
+	deviceConfig: DeviceConfigHook[];
+}
+
+const STATE_ENDPOINT_HOOKS: HookRegistry = {
+	deviceConfig: [],
+};
+
+export function registerHook<Key extends keyof HookRegistry>(
+	name: Key,
+	fn: typeof STATE_ENDPOINT_HOOKS[Key][number],
+) {
+	STATE_ENDPOINT_HOOKS[name].push(fn);
+}
+
+function runHook<Key extends keyof HookRegistry>(
+	name: Key,
+	args: Parameters<typeof STATE_ENDPOINT_HOOKS[Key][number]>,
+) {
+	STATE_ENDPOINT_HOOKS[name].forEach(hook =>
+		hook(...(args as Parameters<typeof hook>)),
```

Comment:
Is this `as` actually necessary?

Reply:
It doesn't work without this. It was your suggestion actually, my initial implementation was doing `hook.apply` which would loose type info but didn't type coercion.

Reply:
Yeah, I do prefer the explicit cast because it at least makes it clear, but wanted to check